compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,292 @@
+
+    int64_t innermostDimSize = sourceType.getShape().back();
+    int64_t transferSizeBits = innermostDimSize * elementBits;
+    LDBG() << "  Source innermost dimension size: " << innermostDimSize;
+    LDBG() << "  Transfer size in bits: " << transferSizeBits;
+
+    std::optional<int64_t> subgroupSize =
+        getSubgroupSize(dmaOp->getParentOfType<FunctionOpInterface>());
+    if (!subgroupSize.has_value()) {
+      LDBG() << "FAILED - unable to determine subgroup size";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "unable to determine subgroup size from forall");
+    }
+    LDBG() << "Subgroup size: " << *subgroupSize;
+
+    // Check that transfer size matches one of the target DMA sizes.
+    int64_t transferSizePerLane = transferSizeBits / *subgroupSize;
+    LDBG() << "Transfer size per lane: " << transferSizePerLane << " bits";
+
+    if (!targetDmaSizes.empty() &&
+        !matchesTargetDmaSize(transferSizePerLane, targetDmaSizes)) {
+      LDBG() << "FAILED - transfer size " << transferSizePerLane
+             << " does not match any target DMA size";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "transfer size does not match any target DMA size");
+    }
+    LDBG() << "Transfer size matches target DMA sizes";
+
+    // TODO: Handle indices properly - for now we skip the explicit indices
+    // check The lane parameter is always present but we handle it differently.
+
+    ArrayRef<int64_t> sourceShape = sourceType.getShape();
+    LDBG() << "Source rank: " << sourceShape.size();
+    if (sourceShape.size() < 2) {
+      LDBG() << "FAILED - source rank < 2";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "source must have at least 2 dimensions");
+    }
+
+    int64_t secondInnermostDimSize = sourceShape[sourceShape.size() - 2];
+    LDBG() << "  Source second innermost dimension size: "
+           << secondInnermostDimSize;
+
+    int64_t elementsPerTransfer = innermostDimSize / *subgroupSize;
+
+    auto transferType = VectorType::get({elementsPerTransfer}, elementType);
+
+    // Actually create the GatherToLDS ops to perform the transfer.
+    rewriter.setInsertionPoint(dmaOp);
+
+    TypedValue<IndexType> laneId = dmaOp.getLane();
+
+    Location loc = dmaOp.getLoc();
+    Value laneOffset = arith::MulIOp::create(
+        rewriter, loc, laneId,
+        arith::ConstantIndexOp::create(rewriter, loc, elementsPerTransfer));
+
+    Value zero = arith::ConstantIndexOp::create(rewriter, loc, 0);
+    for (int64_t i = 0; i < secondInnermostDimSize; ++i) {
+      Value iVal = arith::ConstantIndexOp::create(rewriter, loc, i);
+
+      SmallVector<Value> srcIndices;
+      for (int64_t dim = 0; dim < sourceType.getRank() - 2; ++dim) {
+        srcIndices.push_back(zero);
+      }
+      srcIndices.push_back(iVal);
+      srcIndices.push_back(laneOffset);
+
+      SmallVector<Value> dstIndices;
+      for (int64_t dim = 0; dim < destType.getRank() - 2; ++dim) {
```

Comment:
This is `dstIndices(destType.getRank() - 2, zero);` (or maybe I've got it in the wrong order). Heck, `dstIndices(destType.getRank(), zero)` and overwriting the one index of interest